libs/template/schema_test.go

```diff
@@ -0,0 +1,322 @@
+
+	// assert validation failing for floats
+	err = validateType(true, PropertyTypeNumber)
+	assert.ErrorContains(t, err, "expected type float, but value is true")
+	err = validateType("abc", PropertyTypeNumber)
+	assert.ErrorContains(t, err, "expected type float, but value is \"abc\"")
+
+	// assert validation failing for boolean
+	err = validateType(int(1), PropertyTypeBoolean)
+	assert.ErrorContains(t, err, "expected type boolean, but value is 1")
+	err = validateType(float64(1), PropertyTypeBoolean)
+	assert.ErrorContains(t, err, "expected type boolean, but value is 1")
+	err = validateType("abc", PropertyTypeBoolean)
+	assert.ErrorContains(t, err, "expected type boolean, but value is \"abc\"")
+
+	// assert validation failing for string
+	err = validateType(int(1), PropertyTypeString)
+	assert.ErrorContains(t, err, "expected type string, but value is 1")
+	err = validateType(float64(1), PropertyTypeString)
+	assert.ErrorContains(t, err, "expected type string, but value is 1")
+	err = validateType(false, PropertyTypeString)
+	assert.ErrorContains(t, err, "expected type string, but value is false")
+}
+
+func TestTemplateSchemaValidateConfig(t *testing.T) {
+	// define schema for config
+	schemaJson := `{
+			"properties": {
+				"int_val": {
+					"type": "integer"
+				},
+				"float_val": {
+					"type": "number"
+				},
+				"bool_val": {
+					"type": "boolean"
+				},
+				"string_val": {
+					"type": "string"
+				}
+			}
+		}`
+	var schema Schema
+	err := json.Unmarshal([]byte(schemaJson), &schema)
+	require.NoError(t, err)
+
+	// define the config
+	config := map[string]any{
+		"int_val":    1,
+		"float_val":  1.1,
+		"bool_val":   true,
+		"string_val": "abc",
+	}
+
+	err = validateConfigValueTypes(config, &schema)
+	assert.NoError(t, err)
+}
+
+func TestTemplateSchemaValidateConfigFailsForUnknownField(t *testing.T) {
+	// define schema for config
+	schemaJson := `{
+		"properties": {
+			"int_val": {
+				"type": "integer"
+			},
+			"float_val": {
+				"type": "number"
+			},
+			"bool_val": {
+				"type": "boolean"
+			},
+			"string_val": {
+				"type": "string"
+			}
+		}
+	}`
+	var schema Schema
+	err := json.Unmarshal([]byte(schemaJson), &schema)
+	require.NoError(t, err)
+
+	// define the config
+	config := map[string]any{
+		"foo":        1,
+		"float_val":  1.1,
+		"bool_val":   true,
+		"string_val": "abc",
+	}
+
+	err = validateConfigValueTypes(config, &schema)
+	assert.ErrorContains(t, err, "foo is not defined as an input parameter for the template")
+}
+
+func TestTemplateSchemaValidateConfigFailsForWhenIncorrectTypes(t *testing.T) {
+	// define schema for config
+	schemaJson := `{
+		"properties": {
+			"int_val": {
+				"type": "integer"
+			},
+			"float_val": {
+				"type": "number"
+			},
+			"bool_val": {
+				"type": "boolean"
+			},
+			"string_val": {
+				"type": "string"
+			}
+		}
+	}`
+	var schema Schema
+	err := json.Unmarshal([]byte(schemaJson), &schema)
+	require.NoError(t, err)
```

Comment:
This same schema is used a bunch of times. Please move it to a helper function `testSchema() *Schema` or something so it doesn't need to be repeated in each test.

Reply:
done